File: Data Structures & Algorithms/two-integer-sum/submission-3.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> x;
        
        for (int i = 0; i < nums.size(); i++) {
            int item = nums[i];
            int complement = target - item;
            
            auto it = x.find(complement);
            if (it != x.end()) {
                return {it->second, i};
            }
            x[item] = i;
        }
        
        return {};
    }
};

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-3.py
import bisect
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        # okay so at every nums[i], we can either start a new seq or improve 
        # [9 1 4 2 3 3 7]
        # [9]
        # [1 4 7]
        # [1 2 3 3 7]
        # max is the bottom. 
        # we can rather try overwriting the moment we get a better val.
        # put it in correct spot so we can binary search and find where to write
        
        temp = [] 
        for item in nums:
            if not temp or item > temp[-1]:
                temp.append(item) # start one if not temp or extend
                continue
            # this means we need to overwrite at some pos.
            write_index = bisect.bisect_left(temp, item)
            temp[write_index] = item
        
        return len(temp)